### PR TITLE
feat: support customer managed instance configurations

### PR DESCRIPTION
**Pull request by AlexandrTrf**
Adds samples for working with user managed instance configurations for Cloud Spanner.

### CODE REVIEW

**Comment by amanda-tarafa** [comment]
I'll review by mid week next week. Thanks.

**Comment by AlexandrTrf** [comment]
> I hadn't realized this was a draft, @AlexandrTrf can you confirm if this is ready for review or not? Thanks.

@amanda-tarafa this code is ready. But it uses Google.Cloud.Scanner libraries that have not yet been released and now this code will not be able to be buit and the tests will not be passed.

File: spanner/api/Spanner.Samples.Tests/SpannerFixture.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:40]: You need to use this as a name prefix as instance configs are not database specific, so all tests running on the same project will attempt to use the same `custom-nam3` config. Whenver you need to do some crud with a config you need to append a timestamp or a GUID to this prefix, and you'll need to clean configs up on dispose the same way we clean databases, backups and other resources up.

Else all of these tests are guaranteed to be very flaky and/or run into quota issues.

File: spanner/api/Spanner.Samples/ListInstanceConfigs.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/ListInstanceConfigs.cs:30]: nit: Undo this white space change

File: spanner/api/Spanner.Samples/ListInstanceConfigOperations.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/ListInstanceConfigOperations.cs:33]: Please only print 5 elements, the same as in other listing examples, to make sure that we are not hitting the server several times, we were reaching admin quota very easily.

See this commit: https://github.com/GoogleCloudPlatform/dotnet-docs-samples/commit/35baed2aa2a6491523cebfa9d301750088af13b0

File: spanner/api/Spanner.Samples.Tests/ListInstanceConfigOperationsTest.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples.Tests/ListInstanceConfigOperationsTest.cs:33]:
```suggestion
            Assert.NotNull(configs);
```
Same reason as my other comment, we were hitting admin quoata, see https://github.com/GoogleCloudPlatform/dotnet-docs-samples/commit/35baed2aa2a6491523cebfa9d301750088af13b0

**Review by amanda-tarafa — CHANGES_REQUESTED** [review]
I've made some requests that will apply to all samples and/or tests.

**Comment by AlexandrTrf** [comment]
@amanda-tarafa I have made the necessary changes. Could you do a review?

**Comment by jskeet** [comment]
@AlexandrTrf: Amanda is on vacation until next week.

File: spanner/api/Spanner.Samples/ListInstanceConfigsAsync.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/ListInstanceConfigsAsync.cs:24]: Changing to creating the client async is fine, but you cannot change the file name because the existing sample was already included in docs.

Either we keep the file name incosistent (I'm fine with that in this instance) or you coordinate with the Spanner technical writers to change the docs to start using the new file name.

File: spanner/api/Spanner.Samples/ListInstanceConfigsAsync.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/ListInstanceConfigsAsync.cs:30]: If we are making the method Async, I think this might have an async version. Can you please double check?

File: spanner/api/Spanner.Samples/UpdateInstanceConfigAsync.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/UpdateInstanceConfigAsync.cs:52]: You might want to show polling until completion. Also the text printed will be a little confusing. Maybe be more explicit or even throw.

```suggestion
            updateInstanceConfigOperation = updateInstanceConfigOperation.PollUntilCompleted();
            if (updateInstanceConfigOperation.IsFaulted)
            {
                throw updateInstanceConfigOperation.Exception;
            }
            Console.WriteLine("Update Instance Config operation completed successfully.");
```

File: spanner/api/Spanner.Samples/UpdateInstanceConfigAsync.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/UpdateInstanceConfigAsync.cs:26]: You may return the operation here and check that is successful on the test, for instance.

File: spanner/api/Spanner.Samples.Tests/DeleteInstanceConfigAsyncTest.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples.Tests/DeleteInstanceConfigAsyncTest.cs:34]: If the delete config test runs before the update config test, then the update config test won't work, i.e. won't do anything I know it won't fail because you have a try catch but still we won't test anything.

I prefer if you create a temporary config here to use for deletion.

File: spanner/api/Spanner.Samples.Tests/CreateInstanceConfigAsyncTest.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples.Tests/CreateInstanceConfigAsyncTest.cs:38]: Similar comment here, if this test executes last, then update and delete won't be able to do anything.
I think it's fine if we just have independen configs for each test.

Or is there something I'm missing.

File: spanner/api/Spanner.Samples.Tests/InstanceConfigPropertiesTest.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples.Tests/InstanceConfigPropertiesTest.cs:20]: We don't need these tests here, this is a given from Protobuf. You can assume this will always work.

**Review by amanda-tarafa — CHANGES_REQUESTED** [review]

File: spanner/api/Spanner.Samples/ListInstanceConfigsAsync.cs
Comment (AlexandrTrf) [reply spanner/api/Spanner.Samples/ListInstanceConfigsAsync.cs:30]: Here is my suggestion. 
Can I add the System.Linq.Parallel package to the project in order to print only 5 configurations to the console?

```suggestion
                var instanceConfigs = instanceAdminClient.ListInstanceConfigsAsync(projectName);
        
        // We print the first 5 elements for demonstration purposes.
        // You can print all configs in the sequence by removing the call to Take(5).
        // The sequence will lazily fetch elements in pages as needed.
        await foreach(var instanceConfig in instanceConfigs.Take(5))
        {
            Console.WriteLine($"Available leader options for instance config {instanceConfig.InstanceConfigName.InstanceConfigId}:");
            foreach (var leader in instanceConfig.LeaderOptions)
            {
                Console.WriteLine(leader);
            }

            Console.WriteLine($"Available optional replica for instance config {instanceConfig.InstanceConfigName.InstanceConfigId}:");
            foreach (var optionalReplica in instanceConfig.OptionalReplicas)
            {
                Console.WriteLine($"Replica type - {optionalReplica.Type}, default leader location - {optionalReplica.DefaultLeaderLocation}, location - {optionalReplica.Location}");
            }
        }

        return instanceConfigs.ToEnumerable();
```

**Review by amanda-tarafa — COMMENTED** [review]
A couple of notes on the Async change.

Also let me know if/when the rest is done to re-review all. Thanks

File: spanner/api/Spanner.Samples/ListInstanceConfigs.cs
Comment (AlexandrTrf) [reply spanner/api/Spanner.Samples/ListInstanceConfigs.cs:35]: But otherwise I won 't have an extension method .Take for AsyncEnumerable

Comment (amanda-tarafa) [reply spanner/api/Spanner.Samples/ListInstanceConfigs.cs:35]: Ah yes, that's right, sorry for missing that.

Hmmm, then I think we can keep that part of the method sync for now. It will match all out other listing samples and we don't need to add the extra dependency. Once we have sorted out our quota issues in a different way than just printing 5 elements, we can revisit all listing operations.

**Comment by AlexandrTrf** [comment]
> A couple of notes on the Async change.
> 
> Also let me know if/when the rest is done to re-review all. Thanks

Please perform a review of the code

File: spanner/api/Spanner.Samples.Tests/ListInstanceConfigsAsyncTest.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples.Tests/ListInstanceConfigsAsyncTest.cs:15]: The file name change is not needed anymore as this sample remained sync.

File: spanner/api/Spanner.Samples/DeleteInstanceConfigAsync.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/DeleteInstanceConfigAsync.cs:31]: Instead of trying to obtain the config here (this will fail if the config does not exist), I think it's better to wrap the delete operation in a try catch and look for an RpcException with status "NOT_FOUND", which means the config didn't exist.

File: spanner/api/Spanner.Samples/UpdateInstanceConfigAsync.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/UpdateInstanceConfigAsync.cs:32]: I think you don't need to check here. And in particular if you don't have a try catch the check may fail if the config does not exist.

Also, the purpose of the `UpdateMask` is that you don't have to have the full entity to update it. You create a new (blank) one, set the values that you want to update and set the mask accordingly. I think that shows a better use of the API/Library.

You can add a try catch to the updated operation and check for RpcException wit "NOT_FOUND" which means there was no such configuration before.

**Review by amanda-tarafa — CHANGES_REQUESTED** [review]
A couple of changes, but nothing major.

Thanks!

File: spanner/api/Spanner.Samples/CreateInstanceConfigAsync.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/CreateInstanceConfigAsync.cs:68]:
```suggestion
        if (pollResult.IsFaulted)
        {
            throw pollResult.Exception;
        }
```

File: spanner/api/Spanner.Samples/UpdateInstanceConfigAsync.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/UpdateInstanceConfigAsync.cs:53]:
```suggestion
        if (updateInstanceConfigOperation.IsFaulted)
        {
            throw updateInstanceConfigOperation.Exception;
        }
```

**Review by amanda-tarafa — COMMENTED** [review]
Two tiny nits, but otherwise LGTM.

I'll approve once the library with the needed changes is released and we can run the tests, but otherwise his looks good. Thanks!

**Comment by AlexandrTrf** [comment]
> Two tiny nits, but otherwise LGTM.
> 
> I'll approve once the library with the needed changes is released and we can run the tests, but otherwise his looks good. Thanks!

Done

**Comment by amanda-tarafa** [comment]
@AlexandrTrf this looks good, I'm just waiting for the released library to contain the feature so I can run the tests etc. Once the tests are green I'll appove and merge, thanks.

**Comment by amanda-tarafa** [comment]
I've started the tests after updating the PR. If they are green we are ready to merge.

File: spanner/api/Spanner.Samples.Tests/SpannerFixture.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:130]: This won't build, seems like a conflict resolution gone wrong.

**Review by amanda-tarafa — CHANGES_REQUESTED** [review]
@anshuldavid13  The code as it is is not building. I think it's just a conflict resolution on the fixture or a copy paste mistake. The rest looks fine so make no changes to it until we've been able to run the tests.

**Pull request closed without merging**
